plugins/formal-verification/src/org/jetbrains/kotlin/formver/embeddings/ExpEmbedding.kt

```diff
@@ -0,0 +1,187 @@
+/*
+ * Copyright 2010-2023 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.formver.embeddings
+
+import org.jetbrains.kotlin.formver.domains.NullableDomain
+import org.jetbrains.kotlin.formver.domains.UnitDomain
+import org.jetbrains.kotlin.formver.scala.silicon.ast.*
+import org.jetbrains.kotlin.formver.scala.toScalaBigInt
+import viper.silver.ast.DomainType
+import javax.sound.sampled.LineEvent.Type
+
+sealed interface ExpEmbedding {
+    val type: TypeEmbedding
+    val viperExp: Exp
+
+    fun withType(newType: TypeEmbedding): ExpEmbedding =
```

Comment:
I feel like this function is doing more than it should be.  We already know (since the Kotlin code typechecked) that `type` can be converted to `newType`, the question is just how.  I suspect we should be able to implement this without `isSubTypeOf`.